Camera control ui

bvanwijk opened the pull request:
It is now possible to control the camera's in the UI.
The control buttons are only visible if the selected camera supports this options.
You can test the control of a camera with my own camera.
The UI now retrieves the camera's from the backend server, so you have to start this server first.

#59 need still to be fixed.

jochem725 commented on backend/src/main/java/com/benine/backend/camera/ipcameracontrol/IPCamera2.java:
```diff
@@ -0,0 +1,175 @@
+      e.printStackTrace();
+    } catch (ParserConfigurationException e){
+      e.printStackTrace();
+    }
```
I think this catch statements can be merged into a single one, like
`catch (IOException|SAXException|ParseConfigurationException e){`

jochem725 commented on backend/src/main/java/com/benine/backend/camera/ipcameracontrol/IPCamera2.java:
```diff
@@ -0,0 +1,175 @@
+  @Override
+  public int getZoomPosition() throws CameraConnectionException {
+    //Not supported
+    return 50;
```
Wouldn't it be better to return a value like -1 if an action is not supported? So the UI always takes a default value on -1?

Kingdorian replied:
I don't think that would be better. If it throws a cameraconnectionexception the httprequest simply returns a 400 error. In my opinion the UI should then determine what it should display to the user.

bvanwijk replied:
This was just for testing purposes so you could test the slider for the zoom position will be placed on the right value. The IPCamera2 class is going to be removed as soon as we can use the camera's from Polycast.

jochem725 replied:
I agree with @Kingdorian about the exception, that would be more clear than returning -1. 
While this class is just for testing, we may be able to apply it to other camera's.

jochem725 commented on backend/src/main/java/com/benine/backend/camera/ipcameracontrol/IPCamera2.java:
```diff
@@ -0,0 +1,175 @@
+      json.put("tilt", new Double(getPosition().getTilt()));
+      json.put("zoom", new Double(getZoomPosition()));
+      json.put("iris", 20);
+      json.put("focus", 80);
```
Why does this return 20 and 80 for iris and focus?

Kingdorian replied:
If it has no focus and iris you can simply emit adding those to the json.

bvanwijk replied:
This is just for testing purposes if they are not available the buttons will not be visible in the UI

jochem725 commented:
I've added some comments, overall I think this piece of code needs a bit more comments to make things (like the values) more understandable.

Kingdorian commented on backend/src/main/java/com/benine/backend/Main.java:
```diff
@@ -40,9 +41,28 @@ public static void main(String[] args) {
+    cameraController.addCamera(camera2);
+    SimpleCamera camera3 = new SimpleCamera();
+    camera3.setStreamLink("http://131.180.123.51/cgi-bin/nph-zms?mode=jpeg&monitor=4&scale=100&maxfps=6&buffer=100");
+    cameraController.addCamera(camera3);
```
Please get stuff like this from the config, it makes the code way more readable.

Kingdorian commented on backend/src/main/java/com/benine/backend/camera/ipcameracontrol/IPCamera2.java:
```diff
@@ -0,0 +1,175 @@
+  }
+
+  @Override
+  public int getZoomPosition() throws CameraConnectionException {
```
Why return a hardcoded value of 50? If the camera does not support zoom it should not implement zoomingcamera and thus you won't need to implement this method.

Kingdorian commented on backend/src/main/java/com/benine/backend/camera/ipcameracontrol/IPCamera2.java:
```diff
@@ -0,0 +1,175 @@
+  @Override
+  public void moveTo(Position pos, int panSpeed, int tiltSpeed)
+      throws CameraConnectionException {
+    //not supported
```
Do something to indicate this action is not supported.

Kingdorian commented on backend/src/main/java/com/benine/backend/camera/ipcameracontrol/IPCamera2.java:
```diff
@@ -0,0 +1,175 @@
+  }
+
+  @Override
+  public Position getPosition() throws CameraConnectionException {
```
It would be better to do something here to indicate the action is not supported.

Kingdorian commented:
You're indicating this class is just for test purposes. However I think we should adhere to our normal code quality "rules" if we ever want to keep using this class in the future we won't have to rewrite parts.

I think you could've used a (elaborate) mockito mock object to achieve what you tried to achieve by writing the ipcamera2 class with a lot less work.

jochem725 commented on frontend/public/css/style.css:
```diff
@@ -49,6 +49,53 @@ nav .label {
+    -webkit-appearance: none;
+    background-color: #848484;
+    width: 200px;
+    height:55px;
```
There are some indentation issues here.

jochem725 commented on frontend/public/js/control.js:
```diff
@@ -0,0 +1,139 @@
+		
+		// show stream of every camera in the carousel.
+		for (var c in cameras) {
+			camera_div = camera_area.find('#camera_' + place);
```
Those variables are not declared.

jochem725 commented on frontend/public/js/control.js:
```diff
@@ -0,0 +1,139 @@
+	
+	//determine which elements of the UI to show
+	if (cameras[id].zoom == undefined) {
+		$('.zoomslider').hide();
```
Maybe you could do these jQuery calls once and store them in a variable.

jochem725 commented on frontend/public/js/control.js:
```diff
@@ -0,0 +1,139 @@
+	currentcamera = id;
+	// Show the current camera in the editing view.
+	$('#current_camera').find('img').attr("src", cameras[currentcamera].streamlink);
+	camera_title = $('#current_camera').find('.camera_title');
```
This variable is not declared.

jochem725 commented:
I've added some comments to the JS, you can check this using http://jshint.com. Also works excellent with IntelliJ.

bvanwijk commented:
I have checked the JS with jshint.com and fixed the errors.
I replaced the IPCamera2 class with a test class which creates a mocked camera which can be used for testing.
The test class runs the server so you can use the UI to control this camera.
Because this class runs a server I have excluded it from the tests in maven.

jochem725 commented:
@bvanwijk Great! But why exactly should it be excluded from the maven tests?

bvanwijk commented:
It isn't an automated test, it just starts a server and that's it.
Using this you can test the controls from the UI on my test camera.

Kingdorian commented:
Looks very good!

Kingdorian commented on backend/src/test/java/com/benine/backend/CameraControlTest.java:
```diff
@@ -0,0 +1,188 @@
+   * @throws IOException when the server can not be created.
+   */
+  @Test
+  public void ManualTestUI() throws InterruptedException, IOException{
```
When this is merged in develop and the httpUpdate is merged in develop, we'll have to change this to a HttpController which makes this a class (and especially this part) a lot cleaner.

bvanwijk replied:
Yes, I agree with that.
